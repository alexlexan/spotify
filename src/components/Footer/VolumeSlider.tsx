import React, { ChangeEvent } from "react";
import "./Footer.sass";
import VolumeDownIcon from "@material-ui/icons/VolumeDown";
import { Grid, Slider } from "@material-ui/core";
import { useSelector, useDispatch } from "react-redux";
import Sound from "react-sound";
import { stop } from "../../store/actions";
import { AppState } from "../../store/rootReducer";
import { State, Track } from "../../store/reducer";
import { Actions } from "../../store/actions";
import { ThunkDispatch } from "redux-thunk";

type VolumeSliderProps = {
  activeTrack: Track | null;
  playing: "PLAYING" | "STOPPED" | "PAUSED";
};

const VolumeSlider: React.FC<VolumeSliderProps> = ({
  activeTrack,
  playing,
}) => {
  const dispatch: ThunkDispatch<State, unknown, Actions> = useDispatch();
  const volume = useSelector((state: AppState) => state.spotify.volume);
  const oldVolume = React.useRef(volume);

  const finish = () => {
    dispatch(stop());
  };

  const handleSliderChange = (
    _: ChangeEvent<{}>,
    newValue: number | number[]
  ) => {
    dispatch({ type: "VOLUME", volume: newValue as number });
    oldVolume.current = newValue as number;
  };

  const muteVolume = (volume: number, oldVolume: number | number[]) => {
    const newVolume = volume === 0 ? oldVolume : 0;
    dispatch({ type: "VOLUME", volume: newVolume as number });
  };

  return (
    <Grid container spacing={2}>
      <Grid item>
        <VolumeDownIcon onClick={() => muteVolume(volume, oldVolume.current)} />
      </Grid>
      <Grid item xs>
        {activeTrack && (
          <Sound
            url={activeTrack.preview_url || activeTrack.external_urls.spotify}
            playStatus={playing}
            volume={volume}
            onFinishedPlaying={finish}
          />
        )}
        <Slider
          aria-labelledby="continuous-slider"
          value={volume}
          onChange={handleSliderChange}
        />
      </Grid>
    </Grid>
  );
};

export default VolumeSlider;
