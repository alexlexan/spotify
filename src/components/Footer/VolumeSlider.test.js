import React from "react";
import { useSelector } from "react-redux";
import VolumeSlider from "./VolumeSlider";
import Sound from "react-sound";
import { Slider } from "@material-ui/core";
import VolumeDownIcon from "@material-ui/icons/VolumeDown";

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

describe("test VolumeSlier component", () => {
  let component, useRef;
  let activeTrack = { preview_url: "activeTrackUrl" };
  const playing = "PLAYING";

  beforeEach(() => {
    useSelector.mockImplementation((selector) =>
      selector({ spotify: { volume: 50 } })
    );
    useRef = jest.spyOn(React, "useRef");
  });

  afterEach(() => {
    mockDispatch.mockClear();
  });

  it("should render component without props", () => {
    component = shallow(<VolumeSlider />);
    expect(component).toMatchSnapshot();
  });

  it("should render component with props", () => {
    component = shallow(
      <VolumeSlider activeTrack={activeTrack} playing={playing} />
    );
    expect(component).toMatchSnapshot();
  });
  it("in Sound component should dispatch action Stop and test props ", () => {
    let wrapper = shallow(
      <VolumeSlider activeTrack={activeTrack} playing={"PLAYING"} />
    );
    let propsSound = wrapper.find(Sound).props();
    propsSound.onFinishedPlaying();
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch.mock.calls[0]).toEqual([
      {
        type: "STOP",
      },
    ]);
    expect(propsSound.playStatus).toBe(playing);
    expect(propsSound.volume).toBe(50);
    expect(propsSound.url).toBe(activeTrack.preview_url);
  });

  it("in Sound component url spotify ", () => {
    activeTrack = { external_urls: { spotify: "url" } };
    let wrapper = shallow(
      <VolumeSlider activeTrack={activeTrack} playing={"PLAYING"} />
    );
    let propsSound = wrapper.find(Sound).props();
    expect(propsSound.url).toBe("url");
  });

  it("should dispatch Volume action", () => {
    let wrapper = shallow(
      <VolumeSlider activeTrack={activeTrack} playing={"PLAYING"} />
    );
    let volume = 60;
    wrapper.find(Slider).simulate("change", {}, volume);
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch.mock.calls[0]).toEqual([
      { type: "VOLUME", volume: 60 },
    ]);
  });

  it("should dispatch mute action", () => {
    let wrapper = shallow(
      <VolumeSlider activeTrack={activeTrack} playing={"PLAYING"} />
    );
    let vol = 60;
    let oldVolume = 50;
    wrapper.find(VolumeDownIcon).simulate("click", vol, oldVolume);
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch.mock.calls[0]).toEqual([{ type: "VOLUME", volume: 0 }]);
  });

  it("should dispatch mute action 1", () => {
    useSelector.mockImplementation((selector) =>
      selector({ spotify: { volume: 0 } })
    );
    useRef.mockReturnValue({ current: 20 });
    let wrapper = shallow(
      <VolumeSlider activeTrack={activeTrack} playing={"PLAYING"} />
    );

    wrapper.find(VolumeDownIcon).simulate("click");
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch.mock.calls[0]).toEqual([
      { type: "VOLUME", volume: 20 },
    ]);
  });
});
