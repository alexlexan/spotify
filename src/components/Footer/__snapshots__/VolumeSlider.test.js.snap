// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test VolumeSlier component should render component with props 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  spacing={2}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <VolumeDownIcon
      onClick={[Function]}
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={true}
  >
    <Sound
      autoLoad={false}
      loop={false}
      onBufferChange={[Function]}
      onError={[Function]}
      onFinishedPlaying={[Function]}
      onLoad={[Function]}
      onLoading={[Function]}
      onPause={[Function]}
      onPlaying={[Function]}
      onResume={[Function]}
      onStop={[Function]}
      playStatus="PLAYING"
      playbackRate={1}
      url="activeTrackUrl"
      volume={50}
    />
    <WithStyles(ForwardRef(Slider))
      aria-labelledby="continuous-slider"
      onChange={[Function]}
      value={50}
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`test VolumeSlier component should render component without props 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  spacing={2}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <VolumeDownIcon
      onClick={[Function]}
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={true}
  >
    <WithStyles(ForwardRef(Slider))
      aria-labelledby="continuous-slider"
      onChange={[Function]}
      value={50}
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
